bootstrap/src/main/java/io/airlift/bootstrap/Bootstrap.java

```diff
@@ -183,18 +192,29 @@ public Injector initialize()
         else {
             requiredProperties = requiredConfigurationProperties;
         }
-        SortedMap<String, String> properties = new TreeMap<>();
+        Map<String, String> unusedProperties = new TreeMap<>(requiredProperties);
+
+        // combine property sources
+        Map<String, String> properties = new HashMap<>();
         if (optionalConfigurationProperties != null) {
             properties.putAll(optionalConfigurationProperties);
         }
         properties.putAll(requiredProperties);
         properties.putAll(getSystemProperties());
+
+        // replace environment variables in property values
+        properties = replaceEnvironmentVariables(properties, System.getenv(), (key, error) -> {
```

Comment:
It would be nice to have an option to provide (override) the source of environment variables (`Bootstrap::setEnvironmentVarabilesSupplier(Supplier<Map<String,String>>)`)